runtime/kubernetes/namespace.go

```diff
@@ -0,0 +1,116 @@
+package kubernetes
+
+import (
+	"strings"
+
+	"github.com/micro/go-micro/v3/logger"
+	"github.com/micro/go-micro/v3/util/kubernetes/client"
+)
+
+func (k *kubernetes) ensureNamepaceExists(ns string) error {
+	namespace := client.Format(ns)
+	if namespace == client.DefaultNamespace {
+		return nil
+	}
+
+	exist, err := k.namespaceExists(namespace)
+	if err == nil && exist {
+		return nil
+	}
+	if err != nil {
+		if logger.V(logger.WarnLevel, logger.DefaultLogger) {
+			logger.Warnf("Error checking namespace %v exists: %v", namespace, err)
+		}
+		return err
+	}
+
+	if err := k.createNamespace(namespace); err != nil {
+		if logger.V(logger.WarnLevel, logger.DefaultLogger) {
+			logger.Warnf("Error creating namespace %v: %v", namespace, err)
+		}
+		return err
+	}
+
+	return nil
+}
+
+// namespaceExists returns a boolean indicating if a namespace exists
+func (k *kubernetes) namespaceExists(name string) (bool, error) {
+	// populate the cache
+	if k.namespaces == nil {
```

Comment:
Does this need a mutex?

Reply:
also - how does this cache work if we're running multiple runtime instances? Looks like it only gets populated once so if a different runtime instance creates a namespace then it won't show up in the cache?